### PR TITLE
etcd techdoc analysis

### PR DESCRIPTION
**Pull request by dwelsch-esi**
First draft of the etcd analysis and implementation.

Contributes to: https://github.com/cncf/techdocs/issues/191

### CODE REVIEW

File: assessments/0010-etcd/etcd-analysis.md [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+- [Project documentation][proj-doc]
+- [Contributor documentation][contributor-doc]
+- [Website and documentation infrrastructure][website]
+
+Examples of CNCF documentation that demonstrate the analysis criteria are linked from the [Criteria][cncf-doc-criteria] specification. 
+
+
+### Recommendations, requirements, and best practices
+
+Notwithstanding the fact that this analysis measures documentation against CNCF project maturity standards, in most cases there is more than one way to do things. Few recommendations here are meant to be prescriptive. Rather, the recommended implementations represent the reviewers' experience with how apply documentation best practices. In other words, borrowing terminology from the lexicon of [RFCs][rfc-keywords], the changes described here should be understood as "recommended" or "should" at the strongest, and "optional" or "may" in many cases. Any "must" or "required" actions are clearly denoted as such, and pertain to legal requirements such as copyright and licensing issues.
```
Comment (jbogarthyde) [inline assessments/0010-etcd/etcd-analysis.md:72]: This analysis measures documentation against CNCF project maturity standards, and suggests possible improvements. In most cases...

File: assessments/0010-etcd/etcd-analysis.md [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+There is **high-level conceptual** and design content, but it is hard to find. Most of it is in a *Learning* section, which is not a conventional label for a system overview. 
+
+The *Learning* section, right before *Developer guide* in the ToC, is a catch-all for a lot of different types of information:
+
+- An architectural overview (explanations of how etcd's client and resiliency architectures work).
+- Adoption decision information, in *etcd versus other key-value stores*. This information needs updating.
+- An overview of the gRPC-based API.
+- A glossary.
+
+The documentation is not **feature complete**. There are some **undocumented tasks** associated with key features. For example, Kubernetes and Linux installations is not fully documented. The [Authentication][docs-authentication] page is a stub containing only a CLI example.
```
Comment (jbogarthyde) [inline assessments/0010-etcd/etcd-analysis.md:105]: ...are not...

File: assessments/0010-etcd/etcd-analysis.md [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+
+- A Quickstart guide
+- Installation instructions
+- Tutorials
+- How to tune the server cluster
+- Using the CLI
+
+... and other tasks and procedures.
+
+The documentation contains an adequate introduction and overview, but these are buried in the doc and difficult to find. The adoption path needs more focus and to be split out by user role. For example, "Getting started" should be different for a developer setting up toy server vs. an admin setting up a production server.
```
Comment (jbogarthyde) [inline assessments/0010-etcd/etcd-analysis.md:117]: ...a toy server...

File: assessments/0010-etcd/etcd-analysis.md [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+
+The **"happy path"** for etcd is not a simple procedure and, again, varies by user role. However, I believe that there is sufficient documentation of the most common use cases (configure and run an HA server cluster; set and get key-value pairs via an API from an application) to argue that it is documented. Exception: The Kubernetes and Linux installation instructions noted above. 
+
+Task and tutorial content are **clearly named according to user goals**. Using "-ing" verb forms instead of "How to ..." would make headings easier to navigate.
+
+Tutorials contain large animated graphics of command line activity. Theser are distracting and degrade the usability of the instructions.
```
Comment (jbogarthyde) [inline assessments/0010-etcd/etcd-analysis.md:123]: These...

File: assessments/0010-etcd/etcd-analysis.md [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+
+There are references to previous releases in the documentation. Most of these should probably be removed. 
+
+Benchmarks for down-level versions are available in the current documentation. It's not clear why these old benchmarks haven't been removed from the documentation.
+
+There is an example **installation** on the [etcd Play](http://play.etcd.io/install) server that prsents  "an example workflow to install and deploy etcd." This page consists of a form containing parameters (with no explanation of each parameter other than the label) followed by a CLI script that varies depending on the installation type you choose. Presumably you're supposed to run the script to do the install, but there are no clear task instructions at all. It's not clear how this is useful. Novice users will have a hard time with this. Also, why is this information on the demo server at all? Users should be referred to an installation procedure in the documentation. If this represents a separate, automated installation workflow it should be offered as a procedure in the user doc.
```
Comment (jbogarthyde) [inline assessments/0010-etcd/etcd-analysis.md:135]: that presents...

File: assessments/0010-etcd/etcd-analysis.md [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+
+### New user content
+
+There is a single paragraph on the [website][etcd-io] landing page with a "Learn more" link that goes to the current documetation table of contents. Better would be if it went to an overview page that laid out learning paths for different users ("Start here").
```
Comment (jbogarthyde) [inline assessments/0010-etcd/etcd-analysis.md:139]: There is a single paragraph on the [website][etcd-io] landing page with a "Learn more" link that goes to the current documentation table of contents (ToC). It would be better to link to an overview page that laid out learning paths for different users ("Start here").

File: assessments/0010-etcd/etcd-analysis.md [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+
+The **installation** page gives step-by-step instructions for installing from binaries, from source, and using a package manager. Installation as part of a Kubernetes installation is also described on the page, but there is a placeholder under the heading "Installation as part of Kubernetes installation". Linux installation is also "TBD", which seems like a big omission.
+
+Package manager installation is documented (but not recommended) for Linux and MacOS. **Other OSes** are alluded to but not documented. Supported platforms are documented in the [Operations guide](https://etcd.io/docs/v3.5/op-guide/supported-platform/).
+
+The *Quickstart* section lists suggested next steps in a *What's next?* heading. However, there is no such section on the *Installation* page, and in neither location is a clear **workflow path** delineated.
+
+There is no **new user signpost** in the documentation. The closest equivalent is the *Quickstart* section, but the Quickstart does not differentiate among user roles (Developer? Operator? Trying to decide whether to adopt?).
+
```
Comment (jbogarthyde) [inline assessments/0010-etcd/etcd-analysis.md:150]: Evaluator trying...

**Review by jbogarthyde — COMMENTED** [review]
Overall looks great. A few typos and some minor rewrite suggestions.

File: assessments/0008-backstage/backstage-issues.md
Comment (nate-double-u) [inline assessments/0008-backstage/backstage-issues.md:1]: If we're updating Backstage info, we should break that out into another PR

**Comment by thisisobate** [comment]
@dwelsch-esi @nate-double-u I just updated the website section as planned. Feel free to take a look

File: assessments/0010-etcd/etcd-implementation.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+The top-level documentation recommendations for this project are:
+
+- Complete and update instructional documentation:
+    - Describe etcd's user roles (personas).
+    - Document Kubernetes and Linux installation procedures.
+    - Update and clarify quick start and new user workflows.
+    - Document key tasks as procedures rather than tutorials.
+    - Clarify which user roles (personas) each task is for.
+    - Going forward, write release notes for each major and minor release.
```
Comment (jberkus) [inline assessments/0010-etcd/etcd-implementation.md:24]: Etcd does have release notes in the repository.  Here, are you saying also copy those to the user documentation?

File: assessments/0010-etcd/etcd-implementation.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+- Reorganize the documentation.
+    - Move *project* documentation (documentation for OSS contributors) to the etcd GitHub repo. Move *product* documentation (documentation for those who use the etcd software, in any capacity) to the technical documentation in the website repo.
```
Comment (jberkus) [inline assessments/0010-etcd/etcd-implementation.md:26]: What's the benefit of moving contributor documentation to the main repository?  I understand having it in its own section, but aren't there benefits to having it in the more readable & structured format supplied by doc publishing?

File: assessments/0010-etcd/etcd-implementation.md [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+For example, the [How to access etcd][access-tutorial] tutorial should be two separate tasks: "Writing to etcd on the CLI" and "Reading from etcd on the CLI".
+
+Name the tasks using gerund phrases ("-ing" form, like "Reading ..." and "Writing ..."). Eliminate "How to", which clutters the heading with a redundant phrase that the user must skip over. 
+
+## Describe etcd's user roles
+
+In an "Overview" or "Start here" page, outline etcd's user roles or personas. These probably include:
+- *Evaluator*: Someone trying to determine whether etcd is appropriate for their product, project, or organization.
+- *Admin or Operator*: Someonereponsible for setting up and maintaining a production etcd service.
```
Comment (jberkus) [inline assessments/0010-etcd/etcd-implementation.md:47]: It's worth splitting Admin into two sub-roles:  (a) Admin using Etcd embedded in Kubernetes (b) admin using Etcd on its own.

File: assessments/0010-etcd/etcd-implementation.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+Linux installation instructions are missing ("TDB") from the Installation page. Write and test a procedure for this.
+
+## Update quickstart and new user workflows
+
+Write separate "Getting started" workflows for Developer and Operator users.
+
+## Write release notes
+
+Write release notes for each major and minor release going forward. Release notes should include:
```
Comment (jberkus) [inline assessments/0010-etcd/etcd-implementation.md:64]: See comment above

File: assessments/0010-etcd/etcd-issues.md [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+  - Remove information about old software revisions from the current version's docs if it's not necessary for a significant segment of users. This includes upgrade procedures and performance benchmarks.
+
+
+# Complete and update instructional documentation
+
+## Issue: Describe etcd's user roles
+
+In an "Overview" or "Start here" page, outline etcd's user roles or personas, including:
+- *Evaluator*: Someone trying to determine whether etcd is appropriate for their product, project, or organization.
+- *Admin or Operator*: Someonereponsible for setting up and maintaining a production etcd service.
```
Comment (jberkus) [inline assessments/0010-etcd/etcd-issues.md:33]: Same comment as on etcd-implementation

**Review by jberkus — APPROVED** [review]
In general, this is everything we were trying to get done in 2022.  So overall a huge improvement.

I've got a couple questions/comments though, the biggest ones being:

1. Need to understand the rationale of not including contributor docs in the main docs.
2. Role split for admins

File: assessments/0010-etcd/etcd-implementation.md (hunk 9 above)
Comment (dwelsch-esi) [reply assessments/0010-etcd/etcd-implementation.md:26]: Again, it's a best practice designed to separate content for different user roles. By analogy with commercial software, you wouldn't put (presumably internal) developer documentation in your user manuals. I realize that in OSS there's no proprietary information to keep hidden, but there are other reasons to maintain that separation. A big one in my experience is to avoid confusing new users, who don't yet have a good mental model of the product and are relying on the documentation to build that understanding.

That said, in OSS projects the lines are sometimes blurred between contributors and certain users. And there are sometimes good reasons to put contributor docs on the website; your readability example is one. I recommend that If you do put contributor documentation in the tech docs, clearly label it as such and isolate it from the user documentation.

File: assessments/0010-etcd/etcd-implementation.md (hunk 10 above)
Comment (dwelsch-esi) [reply assessments/0010-etcd/etcd-implementation.md:47]: That's an excellent suggestion. I'll need to collaborate with a contributor who understands the differences between the sub-roles in order to update the implementation plan.

File: assessments/0010-etcd/etcd-implementation.md (hunk 11 above)
Comment (dwelsch-esi) [reply assessments/0010-etcd/etcd-implementation.md:64]: Yup, I wasn't aware of the release notes when I wrote this item. We can remove this as an issue.

**Review by dwelsch-esi — COMMENTED** [review]
Addressed @jberkus 's review comments. Excellent questions; I've tried to address them from a tech writer's perspective. Let me know what you think.

**Review by nate-double-u — APPROVED** [review]
Thanks for this @dwelsch-esi, and thanks for the reviews everyone!
